### ISSUE TITLE
Enable SimpleDRM in the kernel config

### ISSUE THREAD

**Issue by ilya-fedin** [NixOS/nixpkgs#155533]
### Describe the bug
This is more a feature request, but I can't find a template to ask a feature change in build configuration for a package.
There's a new DRM driver in the kernel that replaces the framebuffer in case there's no vendor-provided driver. I was about to create a PR, but then I found a comment in common-config.nix that I should test my change by building kernels, this is not I can do though, so I create this issue and kindly ask to test this and apply if possible.
```diff
diff --git a/pkgs/os-specific/linux/kernel/common-config.nix b/pkgs/os-specific/linux/kernel/common-config.nix
index 93c6b390eb2..4fc63580fa0 100644
--- a/pkgs/os-specific/linux/kernel/common-config.nix
+++ b/pkgs/os-specific/linux/kernel/common-config.nix
@@ -237,7 +237,8 @@ let
 
     fb = {
       FB                  = yes;
-      FB_EFI              = yes;
+      FB_SIMPLE           = whenAtLeast "5.15" no;
+      FB_EFI              = mkMerge [ (whenOlder "5.15" yes) (whenAtLeast "5.15" no) ];
       FB_NVIDIA_I2C       = yes; # Enable DDC Support
       FB_RIVA_I2C         = yes;
       FB_ATY_CT           = yes; # Mach64 CT/VT/GT/LT (incl. 3D RAGE) support
@@ -252,11 +253,12 @@ let
       FRAMEBUFFER_CONSOLE_DEFERRED_TAKEOVER = whenAtLeast "4.19" yes;
       FRAMEBUFFER_CONSOLE_ROTATION = yes;
       FB_GEODE            = mkIf (stdenv.hostPlatform.system == "i686-linux") yes;
-      # On 5.14 this conflicts with FB_SIMPLE.
-      DRM_SIMPLEDRM = whenAtLeast "5.14" no;
+      SYSFB_SIMPLEFB      = whenAtLeast "5.15" yes;
     };
 
     video = {
+      DRM                    = whenAtLeast "5.15" yes;
+      DRM_SIMPLEDRM          = mkMerge [ (whenBetween "5.14" "5.15" no) (whenAtLeast "5.15" yes) ];
       # Allow specifying custom EDID on the kernel command line
       DRM_LOAD_EDID_FIRMWARE = yes;
       VGA_SWITCHEROO         = yes; # Hybrid graphics support
@@ -926,7 +928,8 @@ let
       # Many ARM SBCs hand off a pre-configured framebuffer.
       # This always can can be replaced by the actual native driver.
       # Keeping it a built-in ensures it will be used if possible.
-      FB_SIMPLE = yes;
+      # Since 5.15, SimpleDRM replaces the simple framebuffer.
+      FB_SIMPLE = whenOlder "5.15" yes;
 
     } // optionalAttrs (versionAtLeast version "5.4" && (stdenv.hostPlatform.system == "x86_64-linux" || stdenv.hostPlatform.system == "aarch64-linux")) {
       # Required for various hardware features on Chrome OS devices
```

### Steps To Reproduce
1. Boot the system

### Expected behavior
SimpleDRM is active in the kernel and used rather than old framebuffer drivers

### Notify maintainers
@TredwellGit @ma27 @nequissimus @alyssais

### Metadata
```console
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.11-zen1, NixOS, 22.05 (Quokka)`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.6.0pre20211217_6e6e998`
 - channels(root): `"nixos-22.05pre338578.a2e281f5770"`
 - nixpkgs: `/etc/nixpkgs`
```

**Comment by samueldr** [comment]
See #142015 for some context on the decisions taken.

**Comment by ilya-fedin** [comment]
:thinking: since 21.11 is already released, could this be activated? is there any reason why CONFIG_DRM couldn't be built-in (it should eventually replace CONFIG_FB that is built-in, so it's logical to have DRM built-in)?

**Comment by samueldr** [comment]
We can't break video during early boot, at best it's rude, at worst it can strand users. Note that starting with 5.15, it is used for EFI and VESA framebuffers too.

The current framebuffer options are *all built-ins*. Moving to modules is a big change. Changing them to built-ins was deemed unacceptable by other contributors in the previous discussion. (¯\\\_(ツ)\_/¯)

Additionally, there are a few deficiencies with the NixOS modules system as it is currently implemented for stage1 (initramfs), that are making this harder than it needs to be.

If built as a module, it will add an additional module that will need to be present in the default set of stage-1 modules, which in turn is problematic because `boot.initrd.availableKernelModules` is a list of modules that *must* be present when building stage-1 or else the build fails. The module could be added conditionally on the kernel version, but this adds a new module that will cause a build failure for custom kernel configurations. I'm not sure this can move forward as modules without `boot.initrd.availableKernelModules` being redone to be less strict.

* * *

I personally think `DRM_SIMPLEDRM` should indeed be enabled as soon as we can, but the changes tested thoroughly first.

I also think it might be better as a built-in. Not because it simplifies some things, but because it means the display will be available ASAP to show the current boot status. Otherwise early boot issues may get needlessly hard to diagnose, where currently they are not.

**Comment by ilya-fedin** [comment]
> Changing them to built-ins was deemed unacceptable by other contributors in the previous discussion. (¯\_(ツ)_/¯)

Where though? :thinking:
The only comment in #142015 about that I see is
> IMHO, this all should be M and made configurable.
> Boards without any graphical output (except serial) don't need to have all the drm stuff in their initrd.

But these boards already have framebuffer modules built-in the kernel and having DRM built-in shouldn't hurt them

**Comment by Ma27** [comment]
> but this adds a new module that will cause a build failure for custom kernel configurations

Is that because you'd need to decide on DRM_SIMPLEDRM vs FB_SIMPLE somehow?

I don't know why, but it's kinda hard to follow through this discussion.
What I understand is that making DRM_SIMPLEDRM a builtin is preferred over making this a module, correct (because then we'd have to make sure we only load the module if fb_simple is not a builtin).

What I fail to read out of the previous discussion is: will we actually break hardware when replacing the builtin FB_SIMPLEFB with DRM?

And if not, what's exactly a reason not do to that considering that this is what upstream prefers us to do (from what I understood so far at least)?

**Comment by SuperSandro2000** [comment]
Fixed by above PRs
